Changes the position of the help icon

phantomjinx opened the pull request:
This seeks to address the issues previously discussed. Please review and get back to me with thoughts.

ScreenGrab: http://phantomjinx.co.uk/stuff/ds-summary-help-icon-move.png

* dataservice-main.html
 * Adds the page-help-icon to the data-service main container and uses an
   absolute position to enforce its right-hand side position.
 * Only 1 entry of the page-help-icon helps keep it consistent across all
   the pages
 * Separates the title from the help icon but includes the title here for
   ensuring each page is again consistent

* dataservicePageController
 * When getting the page title, if there is a custom title then use that,
   otherwise return the standard title. This allows the existing behaviour
   of appending the dataservice id to the end of the title
 * On changing the page, delete the custom title

* dsPageService
 * Adds standard helpIds into page metadata
 * Provides setter for custom helpIds should it be required
 * Provides setter for custom title should it be required

* svcSourceCloneController.js
* dsCloneController.js
* dsEditController.js
* dsTestController.js
 * Adds a custom title on loading of the page

* HelpService
 * Slightly off-topic to rest of commit but avoids an infinite loop. If a
   page is not found then the getHelpPageUrl() calls itself with the
   PAGE-NOT-FOUND key. Unfortunately, this is not found so the function is
   stuck infinitely calling itself.

tejones commented:
@phantomjinx much easier to maintain, however does this allow for fine grained help? The current implementation allows for div level help icons, if we want to go that route. I guess we could still add an individual help icon for those cases though. Also, adding to main feels like a global help vs contextual. Thoughts @mdrillin @blafond @elvisisking?

phantomjinx commented:
@tejones 
So it acts as page help at the moment but I have put in an API function in the DSPageService for setting a customHelpId in the same manner as customTitle. Idea being that 'getHelpId()' would return the custom version before the standard version. This would allow for more fine-grained. Clicking buttons / widgets on the page would have to call DSPageService.setCustomHelpId() but they would've had to do something similar with the page-icon in the page so I don't think that should be a problem. Need to be careful that the customHelpId is reset on page load, in same manner as customTitle so I think that should be ok.

Probably best to conjure up a customHelpId we can test this on to see how it would fly?

tejones commented:
@phantomjinx would still like the opinion of @blafond @mdrillin @elvisisking as to whether or not it makes sense to have contextual help at the top of the page. Still feels like global help (which btw.. might be a good idea??). Maybe this is a UX team question?

Either way.. the current implementation already resolves what you are trying to accomplish for page level/fine-grained contextual help, so I'm not convinced it makes sense to go down this path. Waiting for others to weight in though.

phantomjinx commented:
@tejones 

It doesn't change any functionality. Its focus is sorting out the placement of the icon. So what I have explained is the method of achieving it without having to add the page-help-icon directive in each page

elvisisking commented:
Since the widget was designed to be a **page help** widget, I don't have an issue with keeping the title as part of the widget; especially if we are going to have a title on every page (which now you seem to be advocating also @phantomjinx). The current widget can be placed as many times on the page, with or without titles, as as desired; and each different location could have its own help page. I definitely knew the current widget needed some CSS magic and @phantomjinx with his proposed changes has definitely addressed that.

The proposed changes move the logic that pieces the title and icon together out of the widget and into `dataservice-main.html`. So I really don't see the point of doing that. With that said, I think not having to put the widget in each page is really nice.

To summarize:

- I like the fact the title was in the widget
- I think the icon should align itself horizontally with the title (this might be what @tejones was referring too when he talked about global help)
- I like the fact that displaying the topmost page help title/icon logic has moved into `dataservice-main.html` and out of each page.
- I think having the ability to display more than one help title/icon on a page is something we need (especially for pages with multiple, complicated contexts). I'm not sure driving what help is displayed by setting help context IDs based on control interaction will always work. For instance, the user might be interested in help for a certain section of the page but not have interacted with any of the controls yet. And on very long pages, would the user have to interact with a control and then scroll all the way back up to hit the help icon?

phantomjinx commented:
@elvisisking

Sorry, am not advocating a title on each page. More concerned with showing off how to make title formatting consistent across pages. Easy to avoid displaying those titles that are not required, eg. dashboard. Can add that later if required.

The specific widget icon is too large to be included in multiple locations on the page. The patternfly 'i' icon is more appropriate so would probably have to add a parameter to select the alternative icon.

The requirement to place the page-help widget in multiple locations reinforces the need to separate it from the page title and icon.

Contextual help-icons are a different beast since they will be embedded in the page and can provide their own context id. The purpose of the customHelpId property and function simply determines the content displayed when the help icon at the top-right is called. Examples would be the Export Page where either git or file is selected; that state becomes part of the contextual state of the page so a more custom help page would be appropriate.

blafond commented:
I see benefits for going either direction with these title-bar-help-button approaches, unfortunately :(

I do think that we need a solution to provided context help for multiple items/selections/focused panels within a page.  Context page may contain ALL the HTML for that full page, but what is in focus would drive the scroll/hilite that particular section within the HTML page.  I agree that "?" icon is too large, so can we add a smaller "?"  The "i" is field level, so I wouldn't want to mess that up as a standard.

blafond commented:
Let's go with this change and revisit when UX team provides more feedback.

Also logged https://issues.jboss.org/browse/TEIIDTOOLS-53 to address additional context help links/icons within a UI page (i.e. other than title bar)

mdrillin commented:
- I'd advocate for having having a single slide out for a page.  The context is the page.  So no need to put multiple slide out icons within a page.  If more fine-grained help is needed for the page then the 'i' context icon could be embedded at the appropriate place in the page.
- I agree with Ted that putting the '?' at the very top of the page makes it feel like global help, not page specific help.
- I like the way bluejeans did their slide out.  There is an arrow indicator centered vertically on the page at the very right side of the screen.

phantomjinx committed:
Changes the position of the help icon

* dataservice-main.html
 * Adds the page-help-icon to the data-service main container and uses an
   absolute position to enforce its right-hand side position.
 * Only 1 entry of the page-help-icon helps keep it consistent across all
   the pages
 * Separates the title from the help icon but includes the title here for
   ensuring each page is again consistent

* dataservicePageController
 * When getting the page title, if there is a custom title then use that,
   otherwise return the standard title. This allows the existing behaviour
   of appending the dataservice id to the end of the title
 * On changing the page, delete the custom title

* dsPageService
 * Adds standard helpIds into page metadata
 * Provides setter for custom helpIds should it be required
 * Provides setter for custom title should it be required

* svcSourceCloneController.js
* dsCloneController.js
* dsEditController.js
* dsTestController.js
 * Adds a custom title on loading of the page

* HelpService
 * Avoids an infinite loop. If a page is not found then the
   getHelpPageUrl() calls itself with the PAGE-NOT-FOUND key.
   Unfortunately, this is not found so the function is stuck infinitely
   calling itself.

phantomjinx commented:
New version of commit pushed encompassing changes as discussed.

ScreenGrab: http://phantomjinx.co.uk/stuff/ds-summary-help-icon-move-2.png
